pkg/cloudprovider/vsphere/loadbalancer/cleanup.go

```diff
@@ -164,5 +165,13 @@ func (p *lbProvider) CleanupServices(clusterName string, validServices map[types
 			}
 		}
 	}
+
+	// check for orphan unmanaged load balancer service if there are no virtual servers and flag ensureLBServiceDeleted == true
+	if len(lbs) == 0 && ensureLBServiceDeleted {
```

Comment:
Finalizer support should address this use-case and we should pick up finalizer logic in the service controller (see https://github.com/kubernetes/kubernetes/pull/85023) once we bump our kubernetes depedencies to v1.18. You think we should wait for that or add custom logic now?

Reply:
In other words, I _think_ this should be fixed automatically when we release v1.18.0 but can't say without testing it fully

Reply:
We saw this problem only very rarely. So I agree, the finalizer should fix the problem in most cases, but it will be hard to prove it by testing.
And there is still a minor chance, that something goes wrong. The LB is deleted, as soon as the last virtual server has been deleted. If something went wrong with deleting the LB service, it will not be tried a second time without this custom logic.

Reply:
> The LB is deleted, as soon as the last virtual server is being deleted. If something went wrong with deleting the LB service, it will not be tried a second time without this custom logic.

So even if the Service deletion is blocked by the finalizer, there's still a corner case where failure to delete an LB Service will leave it dangling as well? Wouldn't deletion of the LB Service also be gated by the finalizer?

Reply:
A K8s service is representated by one or multiple virtual server. With the finalizer you can only assure that all related virtual servers are deleted. But if the LB service is managed, it has to be deleted after the last virtual server has gone. You would need some kind of a transaction to enforce the deletion of LB service with the last virtual server. 
Here an example, how the LB service may leak without this check:
1. The `EnsureLoadBalancerDeleted` method is called for the last service.
  a) The last virtual server is deleted successfully.
  b) Deleting the LB service fails for some reason.
  c) The finalizer is *not* removed.
2. The `EnsureLoadBalancerDeleted` method is called for the last service a second time
  a) All virtual servers have already been deleted, nothing to do
  b) The finalizer is removed